pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -742,21 +753,33 @@ func (r *reconciler) update(lastRealized *lastRealized, newRule *CompletedRule,
 	// As rule identifier is calculated from the rule's content, the update can
 	// only happen to Group members.
 	if newRule.Direction == v1beta2.DirectionIn {
+		isRuleAppliedToService := isRuleAppliedToService(newRule.TargetMembers)
 		from1 := groupMembersToOFAddresses(newRule.FromAddresses)
-		from2 := ipBlocksToOFAddresses(newRule.From.IPBlocks, r.ipv4Enabled, r.ipv6Enabled)
+		from2 := ipBlocksToOFAddresses(newRule.From.IPBlocks, r.ipv4Enabled, r.ipv6Enabled, isRuleAppliedToService)
 		addedFrom := ipsToOFAddresses(newRule.FromAddresses.IPDifference(lastRealized.FromAddresses))
 		deletedFrom := ipsToOFAddresses(lastRealized.FromAddresses.IPDifference(newRule.FromAddresses))
 
 		membersByServicesMap, servicesMap := groupMembersByServices(newRule.Services, newRule.TargetMembers)
 		for svcKey, members := range membersByServicesMap {
 			newOFPorts := r.getOFPorts(members)
+			newGroupIDAddressSet := sets.NewInt64()
 			ofID, exists := lastRealized.ofIDs[svcKey]
+			toAddresses := make([]types.Address, 0)
+			if isRuleAppliedToService {
+				for _, member := range members.Items() {
+					curAddresses, curAddressSet := r.svcRefToOFAddresses(*member.Service)
+					toAddresses = append(toAddresses, curAddresses...)
+					newGroupIDAddressSet = newGroupIDAddressSet.Union(curAddressSet)
+				}
+			} else {
+				toAddresses = ofPortsToOFAddresses(newOFPorts)
+			}
 			// Install a new Openflow rule if this group doesn't exist, otherwise do incremental update.
 			if !exists {
 				ofRule := &types.PolicyRule{
 					Direction:     v1beta2.DirectionIn,
 					From:          append(from1, from2...),
-					To:            ofPortsToOFAddresses(newOFPorts),
+					To:            toAddresses,
```

Comment:
should it record `newGroupIDAddressSet` in this case?

Reply:
Good catch. Done.